tests/integration/repo_test.go

```diff
@@ -256,3 +256,108 @@ func TestViewRepoDirectory(t *testing.T) {
+			assert.Contains(t, readmeContent.Text(), expectedContent)
+		})
+	}
+
+	// viewing the top level
+	check("Home", "/user2/readme-test/", "README.md", "markdown", "The cake is a lie.")
+
+	// viewing different file extensions
+	check("md", "/user2/readme-test/src/branch/master/", "README.md", "markdown", "The cake is a lie.")
+	check("txt", "/user2/readme-test/src/branch/txt/", "README.txt", "plain-text", "My spoon is too big.")
+	check("plain", "/user2/readme-test/src/branch/plain/", "README", "plain-text", "Birken my stocks gee howdy")
+	check("i18n", "/user2/readme-test/src/branch/i18n/", "README.zh.md", "markdown", "蛋糕是一个谎言")
+
+	// viewing different subdirectories
+	check("subdir", "/user2/readme-test/src/branch/subdir/libcake", "README.md", "markdown", "Four pints of sugar.")
+	check("docs-direct", "/user2/readme-test/src/branch/special-subdir-docs/docs/", "README.md", "markdown", "This is in docs/")
+	check("docs", "/user2/readme-test/src/branch/special-subdir-docs/", "docs/README.md", "markdown", "This is in docs/")
+	check(".gitea", "/user2/readme-test/src/branch/special-subdir-.gitea/", ".gitea/README.md", "markdown", "This is in .gitea/")
+	check(".github", "/user2/readme-test/src/branch/special-subdir-.github/", ".github/README.md", "markdown", "This is in .github/")
+
+	// symlinks
+	// symlinks are subtle:
+	// - they should be able to handle going a reasonable number of times up and down in the tree
+	// - they shouldn't get stuck on link cycles
+	// - they should determine the filetype based on the name of the link, not the target
+	check("symlink", "/user2/readme-test/src/branch/symlink/", "README.md", "markdown", "This is in some/other/path")
+	check("symlink-multiple", "/user2/readme-test/src/branch/symlink/some/", "README.txt", "plain-text", "This is in some/other/path")
+	check("symlink-up-and-down", "/user2/readme-test/src/branch/symlink/up/back/down/down", "README.md", "markdown", "It's a me, mario")
+
+	// testing fallback rules
+	// READMEs are searched in this order:
+	// - [README.zh-cn.md, README.zh_cn.md, README.zh.md, README_zh.md, README.md, README.txt, README,
+	//     docs/README.zh-cn.md, docs/README.zh_cn.md, docs/README.zh.md, docs/README_zh.md, docs/README.md, docs/README.txt, docs/README,
+	//    .gitea/README.zh-cn.md, .gitea/README.zh_cn.md, .gitea/README.zh.md, .gitea/README_zh.md, .gitea/README.md, .gitea/README.txt, .gitea/README,
+
+	//     .github/README.zh-cn.md, .github/README.zh_cn.md, .github/README.zh.md, .github/README_zh.md, .github/README.md, .github/README.txt, .github/README]
+	// and a broken/looped symlink counts as not existing at all and should be skipped.
+	// again, this doesn't cover all cases, but it covers a few
+	check("fallback/top", "/user2/readme-test/src/branch/fallbacks/", "README.en.md", "markdown", "This is README.en.md")
+	check("fallback/2", "/user2/readme-test/src/branch/fallbacks2/", "README.md", "markdown", "This is README.md")
```

Comment:
Strange issue: in the Chinese `README.zh.md` case, when it's the only file, the Chinese text renders successfully, but in this case, if I try to use `README.zh.md` when there's also `README.md`, the English text is preferred.

I swapped out `README.zh.md` for `README.en.md` and that successfully renders the localized README. But why does it allow picking the Chinese README when it's the only one but not when there's two of them? My current understanding of this says it should be ignoring `README.zh.md` in both cases?

https://github.com/go-gitea/gitea/blob/64318798fc0562424a3bad78acabc35a20d587e1/routers/web/repo/view.go#L192-L212

Reply:
It depends on your web browser's language?

Reply:
Indeed. But I don't understand yet then how the test with `README.zh.md` is passing.

The tests, as far as I can tell, are hard-coding the language to 'en':

https://github.com/go-gitea/gitea/blob/79acf7acc4f74d0fe071c1748ae80ff51499706f/modules/test/context_tests.go#L97-L99

so it makes sense that `README.en.md` would render, but `README.zh.md` should never, under test. Yet it is.

I haven't looked too much deeper into it yet. I don't think this quirk devalues the rest of the tests I'm offering here.

Reply:
I understand: given a _viewer's_ locale, there are 6 potential README names:

```
["README.{lang}-{country}.md",
 "README.{lang}_{country}.md",
 "README.{lang}.md",
 "README_{lang}.md",
 "README.md",
 "README.txt",
 "README"]
```

This is defined by

https://github.com/go-gitea/gitea/blob/64318798fc0562424a3bad78acabc35a20d587e1/routers/web/repo/view.go#L207

https://github.com/go-gitea/gitea/blob/64318798fc0562424a3bad78acabc35a20d587e1/routers/web/repo/view.go#L72

However, there's a secret 7th fallback case, hidden over here:

https://github.com/go-gitea/gitea/blob/64318798fc0562424a3bad78acabc35a20d587e1/modules/util/path.go#L237-L239

which means that any "README.*" also counts

And if multiple files match "README.*", it picks the _last_ one (lexically):

https://github.com/go-gitea/gitea/blob/64318798fc0562424a3bad78acabc35a20d587e1/routers/web/repo/view.go#L101

--

So what happened is:

- when there's only "README.zh.md", it matches the "README.*" case
- when there's "README.md" and "README.zh.md", it matches the "README.md" case
- when there's "README.md" and "README.en.md", it matches the "README.en.md" case, which is a case that exists because the tests set their locale to "en".

--

I'm not going to bother adding a test for this though. It's just kind of a counterintuitive quirk.